orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go

```diff
@@ -0,0 +1,176 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reindex
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/hashicorp/go-multierror"
+
+	"magma/orc8r/cloud/go/clock"
+	"magma/orc8r/cloud/go/services/state/indexer/metrics"
+	state_types "magma/orc8r/cloud/go/services/state/types"
+	"magma/orc8r/lib/go/util"
+)
+
+// Major TODOs:
+// TODO(reginawang3495): Remove "Desired" from Indexer.Versions once reindexer_queue and queue are removed
+
+// reindexerSingleton runs as though it is a singleton
+type reindexerSingleton struct {
+	Versioner
+	store     Store
+}
+
+// TODO(reginawang3495): to be setup and used in M2 Part C
+const reindexLoopInterval = 5 * time.Second
+
+func NewReindexerSingleton(store Store, versioner Versioner) Reindexer {
+	return &reindexerSingleton{store: store, Versioner: versioner}
+}
+
+func (r *reindexerSingleton) Run(ctx context.Context) {
+	// indexerID being "" means that we are retrieving jobs on all indices
+	const indexerID = ""
+	// TODO(reginawang3495) will support non-nil sendUpdate from Run after removing queue impl
+	var sendUpdate func(string) = nil
+
+	batches := r.getReindexBatches(ctx)
+	for {
+		if isCanceled(ctx) {
+			glog.Warning("State reindexing async job canceled")
+			return
+		}
+
+		err := r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+		if err != nil {
+			clock.Sleep(failedJobSleep)
+		}
+		clock.Sleep(reindexLoopInterval)
+	}
+}
+
+func (r *reindexerSingleton) RunUnsafe(ctx context.Context, indexerID string, sendUpdate func(string)) error {
+	batches := r.getReindexBatches(ctx)
+	return r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+}
+
+// getReindexBatches gets network-segregated reindex batches with capped number of state IDs per batch.
+func (r *reindexerSingleton) getReindexBatches(ctx context.Context) []reindexBatch {
+	var idsByNetwork state_types.IDsByNetwork
+	for {
+		if isCanceled(ctx) {
+			return nil
+		}
+		ids, err := r.store.GetAllIDs()
+		if err != nil {
+			err = wrap(err, ErrDefault, "")
+			glog.Errorf("Failed to get all state IDs for state indexer reindexing, will retry: %v", err)
+			clock.Sleep(failedGetBatchesSleep)
+			continue
+		}
+		idsByNetwork = ids
+		break
+	}
+
+	var current, rest state_types.IDs
+	var batches []reindexBatch
+	for networkID, ids := range idsByNetwork {
+		rest = ids
+		for len(rest) > 0 {
+			count := util.MinInt(numStatesToReindexPerCall, len(rest))
+			current, rest = rest[:count], rest[count:]
+			batches = append(batches, reindexBatch{networkID: networkID, stateIDs: current})
+		}
+	}
+	return batches
+}
+
+func (r *reindexerSingleton) reindexJobs(ctx context.Context, indexerID string, batches []reindexBatch, sendUpdate func(string)) error {
+	jobs, err := r.getJobs(indexerID)
+	if err != nil {
+		return wrap(err, ErrDefault, indexerID)
+	}
+
+	var errs *multierror.Error
```

Comment:
Why is it better? I think it might not be especially it's not clear that it doesn't need to be initialized + I think you need to initialize it if you are not directly using certain methods on it. EG https://github.com/magma/magma/blob/master/orc8r/cloud/go/obsidian/swagger/spec/combine.go#L68 has to stay same (initialized, not declared)

Reply:
I did make a PR on it though! https://github.com/magma/magma/pull/10098